internal/internal.go

```diff
@@ -695,36 +724,131 @@ func getParsedPathAndQuery(path string) ([]interface{}, map[string]interface{},
 	return parsedPathInterface, parsedQueryInterface, nil
 }
 
-func getParsedBody(headers map[string]string, body string) (interface{}, bool, error) {
-	if body == "" {
-		return nil, false, nil
-	}
-
-	if val, ok := headers["content-length"]; ok {
-		cl, err := strconv.ParseInt(val, 10, 64)
-		if err != nil {
-			return nil, false, err
-		}
-
-		if cl != -1 && cl > int64(len(body)) {
-			return nil, true, nil
-		}
-	}
-
+func getParsedBody(headers map[string]string, body string, rawBody []byte, parsedPath []interface{}, protoSet *protoregistry.Files) (interface{}, bool, error) {
 	var data interface{}
 
 	if val, ok := headers["content-type"]; ok {
 		if strings.Contains(val, "application/json") {
+
+			if body == "" {
+				return nil, false, nil
+			}
+
+			if val, ok := headers["content-length"]; ok {
+				cl, err := strconv.ParseInt(val, 10, 64)
+				if err != nil {
+					return nil, false, err
+				}
+				if cl != -1 && cl > int64(len(body)) {
+					return nil, true, nil
+				}
+			}
+
 			err := util.Unmarshal([]byte(body), &data)
 			if err != nil {
 				return nil, false, err
 			}
+		} else if strings.Contains(val, "application/grpc") {
+
+			if protoSet == nil {
+				return nil, false, nil
+			}
+
+			if len(rawBody) == 0 {
+				return nil, false, fmt.Errorf("invalid raw body")
+			}
+			if len(parsedPath) <= 1 {
+				return nil, false, fmt.Errorf("invalid parsed path")
+			}
+
+			known, truncated, err := getGRPCBody(rawBody, parsedPath, &data, protoSet)
+			if err != nil {
+				return nil, false, err
+			}
+			if truncated {
+				return nil, true, nil
+			}
+			if !known {
+				return nil, false, nil
+			}
 		}
 	}
 
 	return data, false, nil
 }
 
+func getGRPCBody(in []byte, parsedPath []interface{}, data interface{}, files *protoregistry.Files) (found, truncated bool, _ error) {
+
+	// the first 5 bytes are part of gRPC framing. We need to remove them to be able to parse
+	// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
+
+	if len(in) < 5 {
+		return false, false, fmt.Errorf("less than 5 bytes")
+	}
+
+	// Can be 0 or 1, 1 indicates thet the payload is compressed.
+	// The method could be looked up in the request headers, and the
+	// request decompressed; but for now, let's skip it.
+	if in[0] != 0 {
```

Comment:
We are simply checking here if the payload is compressed and for now we won't support that.